File: qt_app/src/service/PublicNodeService.cpp
#include "PublicNodeService.hpp"
#include "LogService.hpp"
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <QRegularExpression>
#include <QUrl>

const QString PublicNodeService::STATUS_URL =
    "https://info.qtet.cn/uptime/status/easytier";
const QString PublicNodeService::HEARTBEAT_URL =
    "https://info.qtet.cn/uptime/api/status-page/heartbeat/easytier";
const QStringList PublicNodeService::RELAY_GROUP_NAMES =
    {"社区公共节点",       // "社区公共节点"
     "社区公共节点[海外]"}; // "社区公共节点[海外]"

PublicNodeService* PublicNodeService::instance() {
    static PublicNodeService s_instance;
    return &s_instance;
}

PublicNodeService::PublicNodeService(QObject* parent)
    : QObject(parent)
{
    m_manager = new QNetworkAccessManager(this);
}

void PublicNodeService::fetchNodes(std::function<void(QList<PublicNode>)> callback) {
    auto* reply = m_manager->get(QNetworkRequest(QUrl(STATUS_URL)));

    connect(reply, &QNetworkReply::finished, this, [this, reply, callback]() {
        reply->deleteLater();

        if (reply->error() != QNetworkReply::NoError) {
            LogService::instance()->error(
                QString("fetchNodes HTTP error: %1").arg(reply->errorString()), "PublicNode");
            callback(QList<PublicNode>());
            return;
        }

        QString html = QString::fromUtf8(reply->readAll());

        // 提取 window.preloadData
        QRegularExpression preloadRe(
            R"(window\.preloadData\s*=\s*(\{[\s\S]*?\});\s*</script>)");
        auto match = preloadRe.match(html);
        if (!match.hasMatch()) {
            LogService::instance()->warn("fetchNodes: preloadData not found", "PublicNode");
            callback(QList<PublicNode>());
            return;
        }

        QString jsStr = match.captured(1);
        QString jsonStr = jsLiteralToJson(jsStr);

        QJsonDocument doc = QJsonDocument::fromJson(jsonStr.toUtf8());
        if (!doc.isObject()) {
            callback(QList<PublicNode>());
            return;
        }

        QJsonObject data = doc.object();
        QJsonArray groups = data["publicGroupList"].toArray();

        QList<PublicNode> nodes;

        for (const auto& g : groups) {
            QJsonObject group = g.toObject();
            QString groupName = group["name"].toString();
            if (!RELAY_GROUP_NAMES.contains(groupName)) continue;

            QJsonArray monitors = group["monitorList"].toArray();
            for (const auto& m : monitors) {
                QJsonObject mObj = m.toObject();
                int id = mObj["id"].toInt();
                QString rawName = mObj["name"].toString();
                QString type = mObj["type"].toString("port");

                auto [serverUrl, desc] = parseNodeName(rawName);
                if (serverUrl.isEmpty()) continue;

                PublicNode node;
                node.id = id;
                node.name = rawName;
                node.serverUrl = resolveUrl(serverUrl);
                node.description = desc;
                node.type = type;
                node.group = groupName;
                nodes.append(node);
            }
        }

        LogService::instance()->info(
            QString("Fetched %1 public nodes").arg(nodes.size()), "PublicNode");
        callback(nodes);
    });
}

void PublicNodeService::attachHeartbeat(QList<PublicNode>& nodes,
                                         std::function<void()> callback) {
    if (nodes.isEmpty()) {
        callback();
        return;
    }

    auto* reply = m_manager->get(QNetworkRequest(QUrl(HEARTBEAT_URL)));

    connect(reply, &QNetworkReply::finished, this, [this, reply, &nodes, callback]() {
        reply->deleteLater();

        if (reply->error() != QNetworkReply::NoError) {
            LogService::instance()->warn(
                QString("heartbeat HTTP error: %1").arg(reply->errorString()), "PublicNode");
            callback();
            return;
        }

        QString body = QString::fromUtf8(reply->readAll());
        QJsonDocument doc = QJsonDocument::fromJson(body.toUtf8());
        if (!doc.isObject()) { callback(); return; }

        QJsonObject data = doc.object();
        QJsonObject heartbeatList = data["heartbeatList"].toObject();

        for (auto& node : nodes) {
            QString key = QString::number(node.id);
            if (!heartbeatList.contains(key)) continue;

            QJsonArray entries = heartbeatList[key].toArray();
            if (entries.isEmpty()) continue;

            QJsonObject last = entries.last().toObject();
            node.status = last["status"].toInt(0);
            if (last.contains("ping"))
                node.ping = last["ping"].toInt(-1);
        }

        callback();
    });
}

QString PublicNodeService::jsLiteralToJson(const QString& input) {
    QString result;
    result.reserve(input.size());
    bool inStr = false;

    for (int i = 0; i < input.size(); ++i) {
        QChar c = input[i];
        if (c == '\\' && inStr && i + 1 < input.size()) {
            result.append(c);
            result.append(input[i + 1]);
            ++i;
        } else if (c == '\'') {
            result.append('"');
            inStr = !inStr;
        } else if (c == '"' && inStr) {
            result.append("\\\"");
        } else {
            result.append(c);
        }
    }

    return result;
}

QPair<QString, QString> PublicNodeService::parseNodeName(const QString& name) {
    // 剥离前缀标签 [xxx]
    QString clean = name.trimmed();
    clean.replace(QRegularExpression(R"(^\[.+?\])"), "").trimmed();

    static QRegularExpression urlRe(R"(^(\w+://[^\s（(]+)[（(](.+)[）)]$)");
    auto match = urlRe.match(clean);
    if (match.hasMatch()) {
        QString url = match.captured(1).trimmed();
        url.replace(QRegularExpression(R"(^\[.+?\])"), "");
        return {url.trimmed(), match.captured(2).trimmed()};
    }
    return {"", ""};
}

QString PublicNodeService::resolveUrl(const QString& originalUrl) {
    QString url = originalUrl;
    url.replace("*", "");
    return url;
}

File: android/app/src/main/java/com/easytier/service/PublicNodeService.kt
package com.easytier.service

import com.easytier.data.PublicNode
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/** 公共节点服务 —— 从 info.qtet.cn 获取社区节点列表和心跳数据 */
object PublicNodeService {
    private const val STATUS_URL = "https://info.qtet.cn/uptime/status/easytier"
    private const val HEARTBEAT_URL = "https://info.qtet.cn/uptime/api/status-page/heartbeat/easytier"

    private val relayGroupNames = listOf("社区公共节点", "社区公共节点[海外]")
    private val urlPattern = Regex("""^(\w+://[^\s（(]+)[（(](.+)[）)]$""")

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    suspend fun fetchNodes(): List<PublicNode> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder().url(STATUS_URL).build()
            val response = client.newCall(request).execute()
            val html = response.body?.string() ?: return@withContext emptyList()

            val preloadMatch = Regex(
                """window\.preloadData\s*=\s*(\{[\s\S]*?\});\s*</script>"""
            ).find(html) ?: return@withContext emptyList()

            val jsStr = preloadMatch.groupValues[1]
            val jsonStr = jsLiteralToJson(jsStr)
            val data = JSONObject(jsonStr)
            val groups = data.optJSONArray("publicGroupList") ?: return@withContext emptyList()

            val nodes = mutableListOf<PublicNode>()

            for (i in 0 until groups.length()) {
                val group = groups.getJSONObject(i)
                val groupName = group.optString("name", "")
                if (groupName !in relayGroupNames) continue

                val monitors = group.optJSONArray("monitorList") ?: continue
                for (j in 0 until monitors.length()) {
                    val m = monitors.getJSONObject(j)
                    val id = m.optInt("id")
                    val rawName = m.optString("name", "")
                    val type = m.optString("type", "port")
                    val (serverUrl, desc) = parseNodeName(rawName)
                    if (serverUrl.isEmpty()) continue

                    nodes.add(PublicNode(
                        id = id,
                        name = rawName,
                        serverUrl = serverUrl,
                        description = desc,
                        type = type,
                        group = groupName
                    ))
                }
            }

            nodes
        } catch (e: Exception) {
            LogService.error("fetch nodes failed: $e", source = "PublicNode")
            emptyList()
        }
    }

    suspend fun attachHeartbeat(nodes: List<PublicNode>) = withContext(Dispatchers.IO) {
        if (nodes.isEmpty()) return@withContext
        try {
            val request = Request.Builder().url(HEARTBEAT_URL).build()
            val response = client.newCall(request).execute()
            val body = response.body?.string() ?: return@withContext
            val data = JSONObject(body)
            val heartbeatList = data.optJSONObject("heartbeatList") ?: return@withContext

            for (node in nodes) {
                val key = node.id.toString()
                val entries = heartbeatList.optJSONArray(key) ?: continue
                if (entries.length() > 0) {
                    val last = entries.getJSONObject(entries.length() - 1)
                    node.status = last.optInt("status", 0)
                    if (last.has("ping")) node.ping = last.optInt("ping")
                }
            }
        } catch (e: Exception) {
            LogService.warn("heartbeat fetch failed: $e", source = "PublicNode")
        }
    }

    private fun parseNodeName(name: String): Pair<String, String> {
        val trimmed = name.trim()
        // 先剥离前缀标签如 [我的]、[海外]，再提取 URL
        val clean = trimmed.replaceFirst(Regex("""^\[.+?]"""), "").trim()
        val match = urlPattern.find(clean)
        if (match != null) {
            var url = match.groupValues[1].trim()
            url = url.replaceFirst(Regex("""^\[.+?]"""), "")
            return Pair(url, match.groupValues[2].trim())
        }
        return Pair("", "")
    }

    /** 将 JavaScript 单引号对象字面量转换为合法 JSON */
    private fun jsLiteralToJson(input: String): String {
        val sb = StringBuilder()
        var inStr = false
        var i = 0

        while (i < input.length) {
            val c = input[i]
            when {
                c == '\\' && inStr && i + 1 < input.length -> {
                    sb.append(c); sb.append(input[i + 1]); i += 2
                }
                c == '\'' -> { sb.append('"'); inStr = !inStr; i++ }
                c == '"' && inStr -> { sb.append("\\\""); i++ }
                else -> { sb.append(c); i++ }
            }
        }
        return sb.toString()
    }
}
